Add Irys for IPFS pinning powered by DeFi

Digital-Mercenaries opened the pull request:
:warning: Current changes are broken!
> Fighting with doc-rot but this description will, hopefully, be updated once I win ;-)

---

## Updates

- **`UploadIpfs.tsx` highlights**
  - [X] input file _should_ be correctly encrypted with uploaded key from web-client and downloaded key from `ZorpStudy.encryptionKey()`
  - [X] once Irys balance checks are sorted, the encrypted data _should_ be uploaded/pinned by Irys
  - [X] when Irys reports something good the CID _should_ be submitted to `ZorpStudy.submitData(cid)`
- **`UploadIpfs.tsx` to-dos** → `zorp-frontend/src/app/zorp/study/submit-data/page.tsx`
  - [x] `zorp-frontend/src/components/features/UploadIpfs.tsx` needs new name
     > figure it be best to warn anyone already testing it out of future churn
- **`UploadIpfs.tsx` and `CreateStudy.tsx` to-dos** → `zorp-frontend/src/app/zorp/factory/create-study/page.tsx`
  - [ ] memory consumption, especially for `opengpg.encrypt`, needs attention
  - [x] Irys check balance in browser needs sorted, for both `UploadIpfs.tsx` and `CreateStudy.tsx`
     > all docs so far found are for NodeJS, or similar, environment(s)
  - [ ] new component configs param, and integration with preexisting contexts could be wise
- **General nice to haves**
  - [ ] documentation for local testing of Irys and/or IPFS would be a good idea

Digital-Mercenaries committed:
Add Irys dot XYZ dependencies as dev-docs dictate

Check the following links for additional development setup instructions;

- https://docs.irys.xyz/build/d/irys-in-the-browser#viem-v2
- https://docs.irys.xyz/build/d/guides/irys-react#step-1-create-a-new-react-project
- https://docs.irys.xyz/build/d/guides/vite

Digital-Mercenaries committed:
Add hopefully correct configs for Anvil dev-net

Digital-Mercenaries committed:
:warning: Track `useProvider` broken state

The "404 Page not found" errors on WAGMI's domain be _extra spicy_

Digital-Mercenaries committed:
Add dependency for CID generation

Digital-Mercenaries committed:
Mostly integrate Irys...  _mostly_

There are oodles of To-Dos that'll need sorted!  But most pressing are;

`openpgp.encrypt` can produce an `Uint8Array` via `format: 'binary'`,
which `sha256.digest` seems happy to encode/hash, but then uploading via
`irysUploadBuilder.uploader.uploadData` doesn't take kindly to that.

`openpgp.encrypt` can produce a `string` via `format: 'armored'`, which
`irysUploadBuilder.uploader.uploadData` is happy enough with, but then
hashing with `sha256.digest` makes TypeScript complain.

These incompatible types, in either case, create a situation where
uploaded file ends up consuming more memory than I'd like.

The hard-coding of `setIrysBalance(0)` needs sorted!  All Irys docs I
could find related to balance queries were for NodeJS, or similar
environments, which ain't applicable to web-browsers.

Digital-Mercenaries committed:
:facepalm: Fix import and adjust variable names

Digital-Mercenaries committed:
Add frontend integration for ZorpFactory

Digital-Mercenaries committed:
Dedicate component to recover GPG key from input

Digital-Mercenaries committed:
Add more Irys SDK dependencies

Digital-Mercenaries committed:
Dedicate component to getting Irys balance

Digital-Mercenaries committed:
Create `ZorpStudy.createStudy` integration page

Digital-Mercenaries committed:
Adjust parameter/state sharing/injection

Digital-Mercenaries committed:
:dizzy: Rename component

Digital-Mercenaries committed:
:fire: Finish `git mv` tracking

Digital-Mercenaries committed:
Track type-hints for Irys

Digital-Mercenaries committed:
Dedicate component for GPG key upload to Irys

Digital-Mercenaries committed:
Add component and adjust state _shape_

Digital-Mercenaries committed:
Correct default `labelText` value

Digital-Mercenaries committed:
:tada: Move final refactoring bits over to page

Digital-Mercenaries committed:
:fire: Remove sketch of features

Digital-Mercenaries committed:
Add untrustworthy config for local testing...

The contract address is likely to change, because order of publishing
and other factors change the transaction bits.

Digital-Mercenaries committed:
Add page to navigation constants

Digital-Mercenaries committed:
Remove unnecessary import and adjust TODO

Digital-Mercenaries committed:
Add fake config for `ZorpStudy` contract on Anvil

Digital-Mercenaries committed:
Add config/constant for Irys

Digital-Mercenaries committed:
Add page for `ZorpStudy.submitData` integration

:warning: Part way through I got the itch to learn the ways of
`useQuery` and thus there are _some_ changes in how this page does its
thang in comparison to `zorp/factory/create-study`

...  Also, as a _gift_ to my future self, the final write contract
functionality still needs moved over.

Digital-Mercenaries committed:
:warning: Adjust type-hint for GPG key

Digital-Mercenaries committed:
Add `ZorpStudy.submitData` integration

Digital-Mercenaries committed:
:fire: Remove sketchy feature implementation

Digital-Mercenaries committed:
Add helper script for Anvil based testing

Digital-Mercenaries committed:
Create list of studies and paginate studies

wkyleg commented on zorp-frontend/src/app/zorp/factory/create-study/page.tsx:
```diff
@@ -0,0 +1,158 @@
+
+			<button
+				className={`zorp_factory_create_study zorp_factory_create_study__button ${className}`}
+				onClick={(event) => {
```
I would break this click event handler into it's own function. You will probably need `useCallback()`

Digital-Mercenaries replied:
Consider it done!  And solid suggestion about `useCallback()`, though considering `useCallback` and `useQuery` both provide caching, and based on my readings `useQuery` offers _more power_ for future re-tries and such, why not `useQuery` on every async callback?

> Side note about previous implementation(s); generally I try to avoid premature abstraction, especially for event handlers the fewer hops in code I do the less likely it be I lose the thread.

wkyleg commented on zorp-frontend/src/app/zorp/study/submit-data/page.tsx:
```diff
@@ -0,0 +1,165 @@
+	const { address, connector, isConnected } = useAccount();
+	const [provider, setProvider] = useState<null | unknown>(null);
+
+	useEffect(() => {
```
`useEffect()` should always come after state variables. These need to be computed in the correct order every time and this is best practice for readability

wkyleg commented on zorp-frontend/src/app/zorp/study/submit-data/page.tsx:
```diff
@@ -0,0 +1,165 @@
+
+					console.log('ZorpStudySubmitData', {event});
+
+					if (!isConnected) {
```
Same here, you will want `useCallback()`

wkyleg commented on zorp-frontend/src/components/features/InputFileToEncryptedMessage.tsx:
```diff
@@ -0,0 +1,98 @@
+	const [inputSubmitDataFile, setInputSubmitDataFile] = useState<null | File>(null);
+	const [message, setMessage] = useState<string>('Info: waiting for public GPG encryption keys and or input file');
+
+	const { data: encryptedMessage } = useQuery({
```
Great use of tanstack here. 

We can make better use of all of the loading/error/succcess states used by WAGMI and tanstack later on

wkyleg commented on zorp-frontend/src/components/features/InputFileToEncryptedMessage.tsx:
```diff
@@ -0,0 +1,98 @@
+			const message = 'Info: attempting to convert input file to ArrayBuffer';
+			console.log('InputFileToEncryptedMessage', {message});
+			setMessage(message);
+			inputSubmitDataFile.arrayBuffer()
```
In general I prefer async/await syntax here with try/catch blocks

Digital-Mercenaries replied:
I'll make the changes, though will caution it may reduce performance (not that that really matters just yet) if what I remember still be correct; TLDR: `await` wraps things in a `Promise` which means a resolved promise gets put at the tail of the queue.for one more round before getting access to the result.

wkyleg commented on zorp-frontend/src/components/features/IrysBalanceGet.tsx:
```diff
@@ -0,0 +1,93 @@
+
+			<button
+				className={`irys_balance irys_balance__button ${className}`}
+				onClick={(event) => {
```
Same here with `useCallback()`

wkyleg commented on zorp-frontend/src/components/features/IrysFetchFileGpgKey.tsx:
```diff
@@ -0,0 +1,83 @@
+		queryFn: () => {
+			const url = `${irysConfig.gatewayUrl.irys}/${cid}`;
+			setMessageFetchEncryptionKey(`Info: attempting to download key from ${url}`);
+			return fetch(url).then((response) => {
```
Great use of tanstack, I would just recommend async/await

wkyleg commented on zorp-frontend/src/components/features/IrysUploadFileGpgKey.tsx:
```diff
@@ -0,0 +1,154 @@
+import * as raw from 'multiformats/codecs/raw';
+import { sha256 } from 'multiformats/hashes/sha2';
+
+import type { BigNumber } from 'bignumber.js';
```
Is it possible to use the Native bignumber type? If not no worries. Sometimes having so many bignumber types can get tricky. Viem/WAGMI natively use the built in type

Digital-Mercenaries replied:
**Short answer**: nope, unfortunately not.

**Slightly longer answer**

That is the type returned by;

```javascript
const balance = await (new WebIrys(webIrysOpts))
    .ready()
    .then((webIrys) => {
        return webIrys.getBalance(address);
    });
```

...  Defined in `node_modules/@irys/sdk/build/cjs/common/irys.d.ts` line `54`, which resolves to;

```typescript
export function BigNumber(n: BigNumber.Value, base?: number): BigNumber;
```

...  in `node_modules/bignumber.js/bignumber.d.ts` on line `1831`

And glancing through the other _few hundred lines_ shows it accepts floats, which ain't compatible with JavaScript® [BigInt][]

[BigInt]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt

wkyleg reviewed (changes requested):
Looks good! Let me know when you're ready to merge.

I think in general we just want slightly better usage of `useCallback()` as that will be less buggy over time with things like form inputs.
https://react.dev/reference/react/useCallback

Also, we generally want to use async/await with error handling. Then later we can also add error boundary components but that doesn't matter right now.

Digital-Mercenaries committed:
Add configs for GitHub Pages deploy... _hopefully_

Digital-Mercenaries committed:
Add GitHub Pages workflow

Digital-Mercenaries committed:
:dizzy: Satisfy linter errors and most warnings

Digital-Mercenaries committed:
Thank you EsLiNt

Digital-Mercenaries committed:
:fire: Remove unnecessary files

Digital-Mercenaries committed:
:facepalm: Fix version for Upload Pages Action

Digital-Mercenaries committed:
Define explicit path for GitHub Pages

Digital-Mercenaries committed:
Correct input

Digital-Mercenaries committed:
Correct what Actions goes wheres

Digital-Mercenaries committed:
Add `/zopr/study/submit-data` to navigation

Digital-Mercenaries committed:
Adjust naming for factory navigation entry

Digital-Mercenaries committed:
Make use of `useCallback` and event handler

Should resolve;

- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976078445
- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976079341

Digital-Mercenaries committed:
`useEffect` should be defined after all `useState`

Should resolve;

- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976079165

Digital-Mercenaries committed:
Prefer `async`/`await` over `Promise` chaining

Should resolve;

- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976080288

Digital-Mercenaries committed:
Use `useCallback`

Should resolve;

- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976080887

Digital-Mercenaries committed:
Prefer `async`/`await` over `Promise` chaining

Should resolve;

- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976081597

Digital-Mercenaries committed:
Make use of `useCallback` for handling `onClick`

Should resolve;

- https://github.com/Elata-Biosciences/zorp/pull/3#discussion_r1976078445

Digital-Mercenaries commented:
> Looks good! Let me know when you're ready to merge.
> 
> I think in general we just want slightly better usage of `useCallback()` as that will be less buggy over time with things like form inputs. https://react.dev/reference/react/useCallback
> 
> Also, we generally want to use async/await with error handling. Then later we can also add error boundary components but that doesn't matter right now.

Thanks for providing easy to follow references to necessary changes!  Made it reasonably easy to _hopefully_ address 'em all.

I believe it be ready for merging, and I'll take the `WiP` prefix off the PR's title :-)

Some follow-up about local testing and other un-checked task-list items;

- I chatted with the Irys team, over on Discord, and at this time local testing ain't supported.  So we'll need to get some public testnet funding to expose any of my implementation mistakes
- Memory consumption, especially for `ZorpStudy.submitData` frontend integration, I predict will be a source of future bug reports.  I don't currently have a good idea on how to combat this though.
- As complexity grows React's `useContext` stuff seems like a good idea, but for now it doesn't seem necessary.

Thangs that were added and other notes of concern;

- GitHub Actions will now automatically run `zorp-contracts` tests, then build the site, then publish it!  The `PAGES_BASE_PATH` in `.github/workflows/github-pages.yaml` and `zorp-frontend/next.config.js` files may need adjusted if y'all decide to use this in production.
- `scripts/zorp-contracts-publish-to-anvil.sh` is showing some strange behavior when running locally via Anvil!  All created study contract addresses are zero-ed and the counter doesn't increment, but unit tests ran with `forge` show expectations are honored?  So I may be missing something, or there may be a bug with one of 'em :shrug:  Either way testing on a _live_ and public dev-net would be a good next step.
